CMS sign digest (WIP)

zuiderkwast opened the pull request:
Add `-digest HEXVALUE` parameter to `openssl cms`, used instead of hashing the content when signing. The option cannot be used with `-nodetached`.

Implementation is not yet working. Test is added but not passing.

Fixes #14686 

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING.md

Other than that, provide a description above this comment if there isn't one already

If this fixes a GitHub issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [ ] tests are added or updated

FdaSilvaYY commented on include/openssl/cms.h.in:
```diff
@@ -96,7 +96,8 @@ CMS_ContentInfo *CMS_ContentInfo_new_ex(OSSL_LIB_CTX *libctx, const char *propq)
 
 BIO *CMS_dataInit(CMS_ContentInfo *cms, BIO *icont);
-int CMS_dataFinal(CMS_ContentInfo *cms, BIO *bio);
+int CMS_dataFinal(CMS_ContentInfo *cms, BIO *bio,
```
Oups ! you cannot change signatures of public API methods.

zuiderkwast replied:
Ok, I will create a CMS_dataFinalDigest. (Didn't know this is public since I didn't see it documented i any pod file.)

FdaSilvaYY replied:
Every file in `include` root folder is part of Public API.
This one is the file you call when you write :
`#include <openssl/cms.h>`

t8m replied:
> Every file in `include` root folder is part of Public API.

Incorrect. Only files in `include/openssl` are the public API.

FdaSilvaYY reviewed (commented):
Apart a bunch of style nits, you need to keep current API.
§You can call your new methods from old API.

zuiderkwast commented:
@FdaSilvaYY thanks for looking! Any clue why it isn't working?

FdaSilvaYY commented:
> 
> 
> @FdaSilvaYY thanks for looking! Any clue why it isn't working?

Two reasons : 
- https://www.openssl.org/policies/cla.html
- First-time contributors need a maintainer to approve running workflows. [Learn more. ](https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks)

Last one is because some newcomers have abuse of GitHub to run some crypto-mining apps.

zuiderkwast commented:
> > @FdaSilvaYY thanks for looking! Any clue why it isn't working?
> 
> Two reasons :
> 
> * https://www.openssl.org/policies/cla.html
> * First-time contributors need a maintainer to approve running workflows. [Learn more. ](https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks)
> 
> 
> Last one is because some newcomers have abuse of GitHub to run some crypto-mining apps.

Oh, right, but it is also not working locally and I need some help debugging. :-) I guess you didn't spot anything obvious in the code at least. I can ask some colleagues for help.

I will mark it "ready for review" when I get it working and have the CLA sent in.

t8m reviewed (commented):
This will require adding documentation for the new public functions.

kaduk commented:
The added docs should probably discuss the security considerations relating to only having an input digest and not the actual content being signed, the need for only using a trusted source of data, etc.

jannelamm commented:
You should check the return value from CMS_final_digest in cms_smime.c:898. ret is set to 0 and never changed.

zuiderkwast commented:
Thanks @jannelamm! I added `ret = 1;` and the test case passes. (It's obvious when you know it.)

Replaced by #15069. I'm closing this and reopening it from the right source organization (legal reasons).
